[PR 1] DEV-16 add WebRTC support

jmaralo opened the pull request:
This PR adds two new hooks that abstract all the WebRTC logic and provides an easy to use interface. There are two hooks: useWebRTC and useSignal. 

The first one -useWebRTC- is the main hook, it receives the address of the media server (and optionally the peer configuration) and handles all the connection logic, it returns the state of the connection and a ref that must be applied to a video component (like the one of useRef).

The second one -useSignal- is an auxiliary hook that handles all the signaling logic with the media server. It connects through a WebSocket and sends and receives events between the two peers. The messages consists of an _event_ and a _payload_.

It currently does not handle reconnects.

jmaralo committed:
DEV-14 add TestingPage component

Necessary code to display the page when navigating to "/testing"

jmaralo committed:
DEV-14 add TestingPage to router

Include the testingRoute in the main router

jmaralo committed:
DEV-14 add ToggleButton component

This is the basic toggle button for the testing page.

It has two states, state changes are notified through a callback.

jmaralo committed:
DEV-16 add useWebRTC hook

The useWebRTC hook creates a new connection with the media server.

A second hook -useSignal- handles connection signaling.

jmaralo commented:
I accidentally added some commits from the previous branch I was working on, please just check both `useWebRTC.ts` and `useSignal.ts` (of course feel free to also take a peak on the other files ;) )

smorsua committed:
Changes

jmaralo committed:
DEV-16 fix hooks

The previous implementation was broken, these changes fix it

Losina24 commented on src/hooks/useWebRTC.ts:
```diff
@@ -0,0 +1,105 @@
+    const remotePeer = useRef<RTCPeerConnection>(null!)
+
+    function onSignal(signal: Signal) {
+        if (signal.event === "answer") {
```
Put the string values ​​in a map above the document and query the map with a get in the function. This is not a good practice

Losina24 commented on src/hooks/useWebRTC.ts:
```diff
@@ -0,0 +1,105 @@
+import { useEffect, useRef, useState } from "react";
+import { Signal, useSignal } from "./useSignal";
+
+export type WebRTCState = "rejected" | RTCPeerConnectionState
```
Magic string antipattern

Losina24 commented on src/pages/TestingPage/loadTestingData.ts:
```diff
@@ -0,0 +1,5 @@
+import { LoaderFunctionArgs } from "react-router-dom";
+
+export const loadTestingData = async (args: LoaderFunctionArgs) => {
+    return null
```
write a TODO comment if it is not implemented yet

Losina24 commented on src/hooks/useWebRTC.ts:
```diff
@@ -0,0 +1,105 @@
+import { useEffect, useRef, useState } from "react";
```
This file contains too much logic

Losina24 commented on src/hooks/useWebRTC.ts:
```diff
@@ -0,0 +1,105 @@
+    }, [])
+
+    useEffect(() => {
+        if (signalState === WebSocket.OPEN) {
```
if (signalState === WebSocket.OPEN) handshake()

This way we reduce the file size and avoid unnecessary indentations.

Losina24 commented on src/hooks/useWebRTC.ts:
```diff
@@ -0,0 +1,105 @@
+    }, [signalState])
+
+    async function onUnload() {
+        closeSignal("leaving")
```
magic string

Losina24 commented on src/hooks/useWebRTC.ts:
```diff
@@ -0,0 +1,105 @@
+        }
+
+        setState(prev => {
+            if (prev === "rejected") {
```
magic string and unnecessary indent

Losina24 commented on src/hooks/useWebRTC.ts:
```diff
@@ -0,0 +1,105 @@
+    }
+
+    async function onIceCandidate(candidate: RTCIceCandidate | null) {
+        if (!candidate) {
```
Use a guard clause to simplify this and make it more readable:

```ts
async function onIceCandidate(candidate: RTCIceCandidate | null) {
    if (candidate) buildAndSendSignal(Events.CANDIDATE, candidate.toJSON())
}
// Events is an enum with all event names stored (it is used to avoid magic strings)

// With this function you can reuse the creation of the object and not have to keep passing an object in parameters to sendSignal (it is done several times)
const buildAndSendSignal = (event: String, payload: String) => sendSignal({event: event, payload: payload})
```

Losina24 commented on src/hooks/useWebRTC.ts:
```diff
@@ -0,0 +1,105 @@
+        videoElement.current.srcObject = stream
+    }
+
+    async function handshake() {
```
a better way to do this:

```ts
async function handshake() {
    const offer = await remotePeer.current.createOffer()
    if (offer.sdp) return sendOfferSignal(offer.sdp)
    sendCloseSignal()
}

const sendOfferSignal = async (sdp: String) => {
    await remotePeer.current.setLocalDescription({type: Events.OFFER, sdp: offer.sdp})
    buildAndSendSignal(Events.OFFER, offer.sdp)
}

const sendCloseSignal = () => buildAndSendSignal(Events.CLOSE, "failed to create offer")

const buildAndSendSignal = (event: String, payload: String) => sendSignal({event: event, payload: payload})
```

Losina24 commented on src/hooks/useSignal.ts:
```diff
@@ -0,0 +1,66 @@
+};
+
+type SignalCandidate = {
+    event: "candidate";
```
Use enums instead of strings

Losina24 commented on src/hooks/useSignal.ts:
```diff
@@ -0,0 +1,66 @@
+    const websocket = useRef<WebSocket>(null!);
+    const [state, setState] = useState<number>(WebSocket.CONNECTING)
+
+    useEffect(() => {
```
useEffect( () => connectSignal(), [])

Losina24 commented on src/hooks/useSignal.ts:
```diff
@@ -0,0 +1,66 @@
+
+        websocket.current.onclose = _ => closeSignal("ok");
+
+        websocket.current.onmessage = ev => {
```
move ev => {...} to a new function

Losina24 commented on src/hooks/useSignal.ts:
```diff
@@ -0,0 +1,66 @@
+    }
+
+    function closeSignal(reason: string) {
+        if (websocket.current.readyState === WebSocket.CLOSING || websocket.current.readyState === WebSocket.CLOSED) {
```
```ts
function closeSignal (reason: String) {
    if (isWebsocketNotEnabled()) return
    websocket.current.send(JSON.stringify({ event: Events.CLOSE, payload: "ok" }))
    websocket.current.close();
}

function isWebsocketNotEnabled(): boolean {
    if (!isWebsocketClosingOrClosed()) return false
    console.log("signal channel closed, reason:", reason)
    return websocket.current !== null
}

function isWebsocketClosingOrClosed(): boolean {
    return websocket.current.readyState === WebSocket.CLOSING || 
        websocket.current.readyState === WebSocket.CLOSED
}
```

Losina24 reviewed (changes requested):
we can improve some things

jmaralo committed:
DEV-16 refactor and improve hoop

The new hook is up to date with the specification.

useSignal is now the SignalChannel class

Losina24 commented on src/hooks/WebRTC/signal.d.ts:
```diff
@@ -0,0 +1,49 @@
+    "keepalive": number,
+}
+
+type SignalCode =
```
use enum

jmaralo replied:
I don't see the advantage of using it here, why do you suggest to use an enum?

Losina24 replied:
First of all, what means 100, 201 or 321 for example. When u will read this code x time in the future, u'll not know what are the meaning of each number. On the other hand, if u need to add a new signal code, if u have a enum, u have all the values centralized in a single place, right know u have to add a new number in every place where u use it. Also, using integers is not safe because when you receive a 312 for example, u won't know if it is a signalcode or any casual number because it is not typed.

Think in a rest api. It is much more comfortable and reusable to use an enum with values: NOT_FOUND(404), ERROR(500), OK(200). Because when you have to check the value in the middle of the code, it will always be much more readable to put a status == Http.OK than a status == 200. Why 200? Why not 421? is it an ID or a result of an operation? Can I pass you the value that I want? Having an enum you are typing and limiting the options of that value, as well as centralizing the point where those values ​​are defined (open/close principle) and making the code much more readable to others.

Losina24 replied:
And the same with what you have above (SignalKinds)

Losina24 replied:
And when do you know you need an enum? You already knew that you needed an enum, you just didn't realize it, that's why you have grouped all those values ​​in a type called SignalCode. There you are hinting that all those 'numbers' are not actually numbers but rather signal codes.

It's just going a little further with that approach. If you want to follow the principle of single responsibility, you will realize that you are loading this class with the responsibility of knowing what values ​​a signalcode should have, wouldn't it be better if there is something (an enum in another file) that contains that responsibility?

jmaralo replied:
Ok, sorry for leaving this hanging and thanks for the feedback! I've seen some arguments against the use of enums in Typescript and I was skeptic to use them.

I'll make the changes on the next commit.

Losina24 replied:
Sure, there are alternatives ways to do that, but in this case, use enum it is better than a type. Why u dont like to use enums?

jmaralo replied:
It is a controversial topic in TypeScript because it adds things during run time, not during compile time. I've seen arguments both [in favour](https://dev.to/dvddpl/whats-the-problem-with-typescript-enums-2okj) and [against](https://dev.to/ivanzm123/dont-use-enums-in-typescript-they-are-very-dangerous-57bh) their use, so I decided to use a union type before I understand better their advantages and disadvantages

Losina24 replied:
It's fine, but in 99% of use cases it's less expensive to create code that is readable and reusable than it is to lose performance. You have to think about what are the important points of your application and about the performance that using one or the other option means. (but if we're using react, I highly doubt we'd have to throttle application performance to the point of not using enums.)
 There is a well-known phrase from a Microsoft engineer who says:
"Make it correct, make it clear, make it concise, make it fast. In that order."

PD: I want to clarify that I'm not saying that you don't have to take efficiency into account, but I'm much more concerned with using sql queries in nested loops than putting some enum hahahahaha

Losina24 reviewed (changes requested):
Waiting to submited comments

jmaralo committed:
DEV-16 switch SignalCode to enum

An enum is better suited and more readable than an union type.

Losina24 commented on src/hooks/WebRTC/signal.d.ts:
```diff
@@ -0,0 +1,48 @@
+    "keepalive": number,
+}
+
+enum SignalCode {
```
Even better if we put it in a separate file so we can use it in other contexts.
